File: dehaze-python/app/service/ai_memory_service.py
"""AI 对话模块 - 长期记忆服务"""

import json
from datetime import datetime
from typing import Any

from sqlalchemy.ext.asyncio import AsyncSession

from app.core.code import ResultCode
from app.core.exceptions import BusinessException
from app.infrastructure.es.ai_memory_index import delete_memory_doc
from app.models.entity.sys_ai_memory import SysAiMemory
from app.models.schema.ai_memory import MemoryCreate, MemoryResult, MemoryUpdate
from app.models.schema.common import PageResult
from app.repository.ai_memory_repository import ai_memory_repository


class AiMemoryService:
    async def list_memories(
        self,
        db: AsyncSession,
        user_id: int,
        page: int,
        size: int,
        memory_type: str | None = None,
        source: str | None = None,
    ) -> PageResult[MemoryResult]:
        memories, total = await ai_memory_repository.list_by_user(
            db, user_id, memory_type, source, page, size
        )
        return PageResult(list=[MemoryResult.model_validate(m) for m in memories], total=total)

    async def list_archived(
        self,
        db: AsyncSession,
        user_id: int,
        page: int,
        size: int,
        memory_type: str | None = None,
    ) -> PageResult[MemoryResult]:
        """归档记忆查看：被遗忘策略归档（archived=1）且未删除的记忆。"""
        memories, total = await ai_memory_repository.list_archived(
            db, user_id, memory_type, page, size
        )
        return PageResult(list=[MemoryResult.model_validate(m) for m in memories], total=total)

    async def create_memory(self, db: AsyncSession, user_id: int, form: MemoryCreate) -> MemoryResult:
        memory = SysAiMemory(
            user_id=user_id,
            memory_type=form.memoryType,
            content=form.content,
            metadata_=form.metadata,
            importance=form.importance,
            source=form.source,
            status=1,
            archived=0,
        )
        memory = await ai_memory_repository.create(db, memory)
        return MemoryResult.model_validate(memory)

    async def update_memory(
        self, db: AsyncSession, memory_id: int, user_id: int, form: MemoryUpdate
    ) -> MemoryResult:
        memory = await ai_memory_repository.get_by_id_and_user(db, memory_id, user_id)
        if not memory:
            raise BusinessException(ResultCode.RESOURCE_NOT_FOUND, "记忆不存在")
        data = form.model_dump(exclude_unset=True)
        for key, value in data.items():
            setattr(memory, key, value)
        await db.flush()
        await db.refresh(memory)
        return MemoryResult.model_validate(memory)

    async def delete_memory(self, db: AsyncSession, memory_id: int, user_id: int) -> None:
        memory = await ai_memory_repository.get_by_id_and_user(db, memory_id, user_id)
        if not memory:
            raise BusinessException(ResultCode.RESOURCE_NOT_FOUND, "记忆不存在")
        await ai_memory_repository.soft_delete_with_time(db, [memory.id])
        # 同步清除 ES 向量索引，避免残留
        await delete_memory_doc(memory.id)

    async def search_memories(
        self, db: AsyncSession, user_id: int, keyword: str, limit: int = 5
    ) -> list[MemoryResult]:
        memories = await ai_memory_repository.search_by_keyword(db, user_id, keyword, limit)
        # 先完成 ORM 实体序列化，再执行 touch（touch 的 UPDATE 会使实体属性过期，
        # 若在其后再访问属性会触发异步懒加载 MissingGreenlet）。
        results = [MemoryResult.model_validate(m) for m in memories]
        for m in memories:
            await ai_memory_repository.touch(db, m.id)
        return results

    async def get_active_memories(
        self, db: AsyncSession, user_id: int, limit: int = 10
    ) -> list[MemoryResult]:
        memories = await ai_memory_repository.get_active_by_user(db, user_id, limit)
        return [MemoryResult.model_validate(m) for m in memories]

    async def batch_clear(
        self,
        db: AsyncSession,
        user_id: int,
        confirm: bool,
        memory_type: str | None = None,
        start: datetime | None = None,
        end: datetime | None = None,
    ) -> int:
        """批量清空记忆（软删 + 记录 delete_time，30 天内可恢复）。

        三种粒度：全部 / 指定类型 / 指定时间范围。需 confirm 二次确认。
        返回受影响条数。
        """
        if not confirm:
            raise BusinessException(ResultCode.PARAM_ERROR, "批量清空记忆为不可逆操作，需二次确认")
        return await ai_memory_repository.batch_clear(db, user_id, memory_type, start, end)

    async def restore_deleted(
        self,
        db: AsyncSession,
        user_id: int,
        confirm: bool,
        memory_type: str | None = None,
        start: datetime | None = None,
        end: datetime | None = None,
    ) -> int:
        """恢复 30 天恢复窗口内的软删记忆（清 deleted 与 delete_time）。"""
        if not confirm:
            raise BusinessException(ResultCode.PARAM_ERROR, "恢复记忆操作需二次确认")
        memories = await ai_memory_repository.list_deleted_for_restore(
            db, user_id, memory_type, start, end
        )
        if not memories:
            return 0
        ids = [m.id for m in memories]
        return await ai_memory_repository.restore_deleted(db, ids)

    async def export_memories(
        self,
        db: AsyncSession,
        user_id: int,
        fmt: str,
    ) -> tuple[str, str]:
        """导出用户全部活跃记忆。

        fmt: json | markdown。返回 (content_type, content)。
        """
        memories = await ai_memory_repository.get_active_by_user(db, user_id, limit=10000)
        if fmt == "markdown":
            lines = ["# 长期记忆导出", ""]
            for m in memories:
                created_at = m.create_time.strftime("%Y-%m-%d %H:%M:%S") if m.create_time else "-"
                lines.extend(
                    [
                        f"## {m.memory_type}（来源：{m.source}）",
                        f"- 内容：{m.content}",
                        f"- 重要性：{m.importance}",
                        f"- 创建时间：{created_at}",
                        "",
                    ]
                )
            return "text/markdown; charset=utf-8", "\n".join(lines)
        records: list[dict[str, Any]] = []
        for m in memories:
            records.append(
                {
                    "id": m.id,
                    "memory_type": m.memory_type,
                    "content": m.content,
                    "metadata": m.metadata_,
                    "source": m.source,
                    "importance": m.importance,
                    "access_count": m.access_count,
                    "created_at": m.create_time.strftime("%Y-%m-%d %H:%M:%S")
                    if m.create_time
                    else None,
                }
            )
        return "application/json; charset=utf-8", json.dumps(
            {"user_id": user_id, "exported_at": datetime.now().isoformat(), "memories": records},
            ensure_ascii=False,
            indent=2,
        )


ai_memory_service = AiMemoryService()

File: dehaze-java/src/main/java/com/pei/dehaze/service/AiMemoryService.java
package com.pei.dehaze.service;

import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import com.baomidou.mybatisplus.core.metadata.IPage;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.pei.dehaze.common.exception.BusinessException;
import com.pei.dehaze.common.result.ResultCode;
import com.pei.dehaze.mapper.SysAiMemoryMapper;
import com.pei.dehaze.model.entity.SysAiMemory;
import com.pei.dehaze.model.form.AiMemoryCreateForm;
import com.pei.dehaze.model.form.AiMemoryUpdateForm;
import com.pei.dehaze.model.vo.AiMemoryVO;
import com.pei.dehaze.service.impl.AiKbIndexClient;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * AI 长期记忆服务。
 *
 * <p>对齐 dehaze-python {@code ai_memory_service}：软删 + 30 天恢复窗口、归档只读视图、
 * 检索命中重激活（access_count+1 / 刷新衰减计时器 / importance+5）、批量清空与恢复需二次确认。
 *
 * @author dehaze
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AiMemoryService {

    private static final int RECOVERY_WINDOW_DAYS = 30;

    private static final DateTimeFormatter TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

    private final SysAiMemoryMapper memoryMapper;

    private final AuditLogService auditLogService;

    private final ObjectMapper objectMapper;

    private final AiKbIndexClient esDocClient;

    /**
     * 记忆导出结果（控制器直接写响应体）
     */
    public record MemoryExport(String filename, String contentType, String content) {
    }

    public IPage<AiMemoryVO> list(Long userId, int pageNum, int pageSize, String memoryType, String source) {
        LambdaQueryWrapper<SysAiMemory> wrapper = new LambdaQueryWrapper<SysAiMemory>()
                .eq(SysAiMemory::getUserId, userId)
                .eq(SysAiMemory::getStatus, 1)
                .eq(SysAiMemory::getArchived, 0)
                .eq(memoryType != null, SysAiMemory::getMemoryType, memoryType)
                .eq(source != null, SysAiMemory::getSource, source)
                .orderByDesc(SysAiMemory::getImportance)
                .orderByDesc(SysAiMemory::getCreateTime);
        Page<SysAiMemory> page = new Page<>(pageNum, pageSize);
        IPage<SysAiMemory> memoryPage = memoryMapper.selectPage(page, wrapper);
        return pageOf(page, memoryPage.getRecords().stream().map(this::toVO).toList(), memoryPage.getTotal());
    }

    public IPage<AiMemoryVO> listArchived(Long userId, int pageNum, int pageSize, String memoryType) {
        LambdaQueryWrapper<SysAiMemory> wrapper = new LambdaQueryWrapper<SysAiMemory>()
                .eq(SysAiMemory::getUserId, userId)
                .eq(SysAiMemory::getArchived, 1)
                .eq(memoryType != null, SysAiMemory::getMemoryType, memoryType)
                .orderByDesc(SysAiMemory::getImportance)
                .orderByDesc(SysAiMemory::getCreateTime);
        Page<SysAiMemory> page = new Page<>(pageNum, pageSize);
        IPage<SysAiMemory> memoryPage = memoryMapper.selectPage(page, wrapper);
        return pageOf(page, memoryPage.getRecords().stream().map(this::toVO).toList(), memoryPage.getTotal());
    }

    @Transactional
    public AiMemoryVO create(Long userId, AiMemoryCreateForm form) {
        SysAiMemory memory = new SysAiMemory();
        memory.setUserId(userId);
        memory.setMemoryType(form.getMemoryType());
        memory.setContent(form.getContent());
        memory.setMetadata(form.getMetadata());
        memory.setImportance(form.getImportance() == null ? 50 : form.getImportance());
        memory.setSource(form.getSource() == null ? "manual" : form.getSource());
        memory.setStatus(1);
        memory.setArchived(0);
        memory.setAccessCount(0);
        memoryMapper.insert(memory);
        return toVO(memory);
    }

    @Transactional
    public AiMemoryVO update(Long memoryId, Long userId, AiMemoryUpdateForm form) {
        SysAiMemory memory = getOwned(memoryId, userId);
        if (form.getContent() != null) {
            memory.setContent(form.getContent());
        }
        if (form.getImportance() != null) {
            memory.setImportance(form.getImportance());
        }
        if (form.getStatus() != null) {
            memory.setStatus(form.getStatus());
        }
        memoryMapper.updateById(memory);
        return toVO(memory);
    }

    @Transactional
    public void delete(Long memoryId, Long userId) {
        SysAiMemory memory = getOwned(memoryId, userId);
        memoryMapper.softDeleteByIds(List.of(memory.getId()), userId);
        // 同步清除 ES 向量文档，否则已删记忆仍被向量检索召回
        esDocClient.deleteMemoryDoc(memory.getId());
    }

    /**
     * 取消归档：恢复注入并刷新衰减计时器。
     *
     * <p>归档是遗忘策略的系统行为，用户侧只提供"取消归档"这一个反向操作（不存在手动归档）；
     * 不刷新 last_accessed_at 时，下次每日归档任务会按旧时点立刻再次归档，用户操作形同无效。
     */
    @Transactional
    public AiMemoryVO unarchive(Long memoryId, Long userId) {
        SysAiMemory memory = getOwned(memoryId, userId);
        if (!Integer.valueOf(1).equals(memory.getArchived())) {
            throw new BusinessException(ResultCode.DATA_STATE_NOT_ALLOW, "该记忆未处于归档状态");
        }
        memory.setArchived(0);
        memory.setLastAccessedAt(LocalDateTime.now());
        memoryMapper.updateById(memory);
        return toVO(memory);
    }

    /**
     * 关键词搜索：命中后重激活（访问计数 +1、刷新衰减计时器、重要性 +5）
     */
    @Transactional
    public List<AiMemoryVO> search(Long userId, String keyword, int limit) {
        List<SysAiMemory> memories = memoryMapper.selectList(new LambdaQueryWrapper<SysAiMemory>()
                .eq(SysAiMemory::getUserId, userId)
                .eq(SysAiMemory::getStatus, 1)
                .eq(SysAiMemory::getArchived, 0)
                .like(SysAiMemory::getContent, keyword)
                .orderByDesc(SysAiMemory::getImportance)
                .last("LIMIT " + limit));
        List<AiMemoryVO> results = memories.stream().map(this::toVO).toList();
        for (SysAiMemory memory : memories) {
            memoryMapper.touch(memory.getId());
        }
        return results;
    }

    @Transactional
    public int batchClear(Long userId, Boolean confirm, String memoryType,
                          LocalDateTime start, LocalDateTime end) {
        if (!Boolean.TRUE.equals(confirm)) {
            throw new BusinessException(ResultCode.PARAM_ERROR, "批量清空记忆为不可逆操作，需二次确认");
        }
        int count = memoryMapper.batchClear(userId, memoryType, start, end, userId);
        Map<String, Object> afterValue = new LinkedHashMap<>();
        afterValue.put("count", count);
        afterValue.put("memory_type", memoryType);
        afterValue.put("start", start == null ? null : start.toString());
        afterValue.put("end", end == null ? null : end.toString());
        AiTxSupport.afterCommit(() -> auditLogService.recordAudit(userId, "ai_memory", userId, "clear",
                "ai_memory", null, afterValue, null, null));
        return count;
    }

    @Transactional
    public int restoreDeleted(Long userId, Boolean confirm, String memoryType,
                              LocalDateTime start, LocalDateTime end) {
        if (!Boolean.TRUE.equals(confirm)) {
            throw new BusinessException(ResultCode.PARAM_ERROR, "恢复记忆操作需二次确认");
        }
        List<SysAiMemory> deleted = memoryMapper.listDeletedForRestore(userId, memoryType, start, end,
                LocalDateTime.now().minusDays(RECOVERY_WINDOW_DAYS));
        if (deleted.isEmpty()) {
            return 0;
        }
        return memoryMapper.restoreByIds(deleted.stream().map(SysAiMemory::getId).toList(), userId);
    }

    /**
     * 导出用户全部活跃记忆（json/markdown）；记忆属敏感个人数据，批量导出必须留痕
     */
    public MemoryExport export(Long userId, String format) {
        List<SysAiMemory> memories = memoryMapper.selectList(new LambdaQueryWrapper<SysAiMemory>()
                .eq(SysAiMemory::getUserId, userId)
                .eq(SysAiMemory::getStatus, 1)
                .eq(SysAiMemory::getArchived, 0)
                .orderByDesc(SysAiMemory::getImportance)
                .orderByDesc(SysAiMemory::getLastAccessedAt)
                .last("LIMIT 10000"));
        Map<String, Object> afterValue = new LinkedHashMap<>();
        afterValue.put("format", format);
        afterValue.put("count", memories.size());
        auditLogService.recordAudit(userId, "ai_memory", userId, "export", "ai_memory", null, afterValue,
                null, null);
        if ("markdown".equals(format)) {
            StringBuilder sb = new StringBuilder("# 长期记忆导出\n\n");
            for (SysAiMemory memory : memories) {
                sb.append("## ").append(memory.getMemoryType())
                        .append("（来源：").append(memory.getSource()).append("）\n")
                        .append("- 内容：").append(memory.getContent()).append("\n")
                        .append("- 重要性：").append(memory.getImportance()).append("\n")
                        .append("- 创建时间：")
                        .append(memory.getCreateTime() == null ? "-" : memory.getCreateTime().format(TIME_FORMAT))
                        .append("\n\n");
            }
            return new MemoryExport("memories.md", "text/markdown; charset=utf-8", sb.toString());
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("user_id", userId);
        payload.put("exported_at", LocalDateTime.now().toString());
        List<Map<String, Object>> records = new ArrayList<>();
        for (SysAiMemory memory : memories) {
            Map<String, Object> item = new LinkedHashMap<>();
            item.put("id", memory.getId());
            item.put("memory_type", memory.getMemoryType());
            item.put("content", memory.getContent());
            item.put("metadata", memory.getMetadata());
            item.put("source", memory.getSource());
            item.put("importance", memory.getImportance());
            item.put("access_count", memory.getAccessCount());
            item.put("created_at",
                    memory.getCreateTime() == null ? null : memory.getCreateTime().format(TIME_FORMAT));
            records.add(item);
        }
        payload.put("memories", records);
        try {
            return new MemoryExport("memories.json", "application/json; charset=utf-8",
                    objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(payload));
        } catch (Exception e) {
            throw new BusinessException(ResultCode.SYSTEM_EXECUTION_ERROR, "记忆导出失败");
        }
    }

    private SysAiMemory getOwned(Long memoryId, Long userId) {
        SysAiMemory memory = memoryMapper.selectOne(new LambdaQueryWrapper<SysAiMemory>()
                .eq(SysAiMemory::getId, memoryId)
                .eq(SysAiMemory::getUserId, userId));
        if (memory == null) {
            throw new BusinessException(ResultCode.RESOURCE_NOT_FOUND, "记忆不存在");
        }
        return memory;
    }

    private AiMemoryVO toVO(SysAiMemory memory) {
        AiMemoryVO vo = new AiMemoryVO();
        vo.setId(memory.getId());
        vo.setUserId(memory.getUserId());
        vo.setMemoryType(memory.getMemoryType());
        vo.setContent(memory.getContent());
        vo.setMetadata(memory.getMetadata());
        vo.setImportance(memory.getImportance());
        vo.setAccessCount(memory.getAccessCount());
        vo.setLastAccessedAt(memory.getLastAccessedAt());
        vo.setSource(memory.getSource());
        vo.setStatus(memory.getStatus());
        vo.setArchived(memory.getArchived());
        vo.setCreateTime(memory.getCreateTime());
        vo.setUpdateTime(memory.getUpdateTime());
        return vo;
    }

    private <T> IPage<T> pageOf(Page<?> page, List<T> records, long total) {
        Page<T> result = new Page<>(page.getCurrent(), page.getSize(), total);
        result.setRecords(new ArrayList<>(records));
        return result;
    }
}
